### PR TITLE
UCP/CORE: Set super request to INVALID and fix the checks in the request purging

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Set super request to INVALID and fix the checks in the request purging.

## Why ?

```
[jazz09:69037:0:69037]      ucp_ep.c:2696 Assertion `req->super_req == ((void *)0)' failed

/labhome/dmitrygla/work_auto/ucx/src/ucp/core/ucp_ep.c: [ ucp_ep_req_purge() ]
      ...
     2693     }
     2694
     2695     if (req->flags & (UCP_REQUEST_FLAG_SEND_AM | UCP_REQUEST_FLAG_SEND_TAG)) {
==>  2696         ucs_assert(req->super_req == NULL);
     2697         ucs_assert(req->send.ep == ucp_ep);
     2698         ucp_request_complete_and_dereg_send(req, status);
     2699     } else if (req->flags & UCP_REQUEST_FLAG_RECV_AM) {

==== backtrace (tid:  69037) ====
 0 0x00000000000418a6 ucp_ep_req_purge()  /labhome/dmitrygla/work_auto/ucx/src/ucp/core/ucp_ep.c:2696
 1 0x0000000000042dc0 ucp_ep_reqs_purge()  /labhome/dmitrygla/work_auto/ucx/src/ucp/core/ucp_ep.c:2740
 2 0x0000000000053129 ucp_worker_iface_err_handle_progress()  /labhome/dmitrygla/work_auto/ucx/src/ucp/core/ucp_worker.c:442
 3 0x0000000000054ee7 ucs_callbackq_slow_proxy()  /labhome/dmitrygla/work_auto/ucx/src/ucs/datastruct/callbackq.c:402
 4 0x000000000004fd7e ucs_callbackq_dispatch()  /labhome/dmitrygla/work_auto/ucx/src/ucs/datastruct/callbackq.h:211
 5 0x000000000005c9c5 uct_worker_progress()  /labhome/dmitrygla/work_auto/ucx/src/uct/api/uct.h:2592
 6 0x00000000004022c0 request_wait()  /labhome/dmitrygla/work_auto/ucx/examples/ucp_client_server.c:260
 7 0x0000000000402321 request_finalize()  /labhome/dmitrygla/work_auto/ucx/examples/ucp_client_server.c:276
 8 0x000000000040271a send_recv_tag()  /labhome/dmitrygla/work_auto/ucx/examples/ucp_client_server.c:384
 9 0x0000000000403047 client_server_communication()  /labhome/dmitrygla/work_auto/ucx/examples/ucp_client_server.c:676
10 0x00000000004034f6 client_server_do_work()  /labhome/dmitrygla/work_auto/ucx/examples/ucp_client_server.c:833
11 0x0000000000403745 run_client()  /labhome/dmitrygla/work_auto/ucx/examples/ucp_client_server.c:954
12 0x00000000004038f3 main()  /labhome/dmitrygla/work_auto/ucx/examples/ucp_client_server.c:1036
13 0x00000000000223d5 __libc_start_main()  ???:0
14 0x0000000000401759 _start()  ???:0
=================================
[jazz09:69037:0:69037] Process frozen, press Enter to attach a debugger...
```
Fixes touching uninitialized `super_req` field of requests.

## How ?

1. Mark each `super_req` as `UCP_REQUEST_INVALID` - it's better than just setting `NULL`, because UCP can check for `super_req == NULL` and expect something.
2. Expect this is to be set instead of `NULL` in `ucp_ep_req_purge()`

### CODE REVIEW

**Comment by dmitrygx** [comment]
@brminich @hoopoepg could you review pls?

**Review by hoopoepg — APPROVED** [review]

File: src/ucp/core/ucp_request.inl [hunk 1]
```diff
@@ -113,6 +115,22 @@
...
+    ucs_assertv(((_req)->super_req != UCP_REQUEST_INVALID) && \
+                !((_req)->flags & UCP_REQUEST_FLAG_CALLBACK), \
+                "req=%p req->super_req=%p (user cb is %s)", \
+                (_req), (_req)->super_req, \
+                ((_req)->flags & UCP_REQUEST_FLAG_CALLBACK) ? "set" : "not set")
+
+
+#define ucp_request_super_is_invalid(_req) \
+    ucs_assertv(((_req)->super_req == UCP_REQUEST_INVALID) || \
+                ((_req)->flags & UCP_REQUEST_FLAG_CALLBACK), \
```
Comment (yosefe) [inline src/ucp/core/ucp_request.inl:128]: how is the CALLBACK flag related here?

File: src/ucp/core/ucp_request.inl [hunk 2]
```diff
@@ -101,6 +102,7 @@
             } \
         } else { \
             __req = ((ucp_request_t*)(_param)->request) - 1; \
+            ucp_request_super_reset(__req); \
```
Comment (yosefe) [inline src/ucp/core/ucp_request.inl:105]: reqeust alloc macros become too bloated
need to add inline functions to take away code from macros:
1. ucp_request_reset_mpool: ucp_request_super_reset, id_check, VALGRIND_MAKE_MEM_DEFINED
2. ucp_request_reset_external: ucp_request_super_reset, id_reset

File: src/ucp/core/ucp_request.inl [hunk 3]
```diff
@@ -113,6 +115,22 @@
                 (_req), (_req)->id, (_id))
 
 
+#define ucp_request_super_is_valid(_req) \
```
Comment (yosefe) [inline src/ucp/core/ucp_request.inl:118]: ucp_request_assert_super_is_valid

**Review by yosefe — CHANGES_REQUESTED** [review]
How does it work since user_data is union with super_req?

File: src/ucp/core/ucp_request.inl (hunk 2 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_request.inl]: done,
but I guess it is better to have `ucp_request_reset()` which is common for both cases and use it directly for resetting external requests and from `ucp_request_reset_mpool()` when allocating from mpool

File: src/ucp/core/ucp_request.inl (hunk 3 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_request.inl]: done

File: src/ucp/core/ucp_request.inl (hunk 1 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_request.inl]: explained in the comment

**Comment by dmitrygx** [comment]
> How does it work since user_data is union with super_req?

we check for `CALLBACK` flag to check whether `user_data` is valid or not.

**Comment by yosefe** [comment]
> we check for `CALLBACK` flag to check whether `user_data` is valid or not.

I don't think this is a good assumption
better add request debug flag for checking super_req is valid - we use flags when field is part of a union

**Comment by dmitrygx** [comment]
> > we check for `CALLBACK` flag to check whether `user_data` is valid or not.
> 
> I don't think this is a good assumption
> better add request debug flag for checking super_req is valid - we use flags when field is part of a union

then we should update all places where we set `super_req`, is it ok?

**Comment by dmitrygx** [comment]
@yosefe could you review pls?

**Comment by yosefe** [comment]
> then we should update all places where we set `super_req`, is it ok?

Yes, should update all places. Can use ucp_request_get_super() inle wrapper to check SUPER flag is set.

**Comment by yosefe** [comment]
> > then we should update all places where we set `super_req`, is it ok?
> 
> Yes, should update all places. Can use ucp_request_get_super() inline wrapper to check SUPER flag is set.

**Comment by dmitrygx** [comment]
> Yes, should update all places. Can use ucp_request_get_super() inle wrapper to check SUPER flag is set.

ok, makes sense. thank you!
shall we also add `ucp_request_set_super()` inline wrapper?

**Comment by yosefe** [comment]
> > Yes, should update all places. Can use ucp_request_get_super() inle wrapper to check SUPER flag is set.
> 
> ok, makes sense. thank you!
> shall we also add `ucp_request_set_super()` inline wrapper?

yes, good idea

**Comment by dmitrygx** [comment]
> > > Yes, should update all places. Can use ucp_request_get_super() inle wrapper to check SUPER flag is set.
> > 
> > 
> > ok, makes sense. thank you!
> > shall we also add `ucp_request_set_super()` inline wrapper?
> 
> yes, good idea

done

File: src/ucp/core/ucp_request.inl [hunk 4]
```diff
@@ -169,12 +166,34 @@
...
+#endif
+}
 
 static UCS_F_ALWAYS_INLINE void ucp_request_id_reset(ucp_request_t *req)
 {
     req->id = UCP_REQUEST_ID_INVALID;
 }
 
+static UCS_F_ALWAYS_INLINE void
+ucp_request_reset_mpool(ucp_request_t *req, ucp_worker_h worker)
```
Comment (brminich) [inline src/ucp/core/ucp_request.inl:182]: maybe `ucp_request_reset_internal`?

Comment (dmitrygx) [reply src/ucp/core/ucp_request.inl]: done

**Comment by dmitrygx** [comment]
@yosefe @brminich your comments were fixed
could you review pls?

File: src/ucp/core/ucp_request.h [hunk 5]
```diff
@@ -26,6 +26,7 @@
 #include <ucp/core/ucp_worker.h>
 
 
+#define UCP_REQUEST_INVALID ((ucp_request_t*)(uintptr_t)SIZE_MAX)
```
Comment (yosefe) [inline src/ucp/core/ucp_request.h:29]: IMO no longer needed since we have the flag

File: src/ucp/core/ucp_request.inl [hunk 6]
```diff
@@ -912,6 +931,26 @@ ucp_request_get_by_id(ucp_worker_h worker, ucs_ptr_map_key_t id,
     return status;
 }
 
+static UCS_F_ALWAYS_INLINE void ucp_request_set_super(ucp_request_t *req,
+                                                      ucp_request_t *super_req)
+{
+    ucs_assertv(!(req->flags & UCP_REQUEST_FLAG_SUPER_VALID) ||
+                (req->super_req == super_req),
```
Comment (yosefe) [inline src/ucp/core/ucp_request.inl:938]: do we have case where we set super_req twice to same thing?

Comment (dmitrygx) [reply src/ucp/core/ucp_request.inl:930]: yes, e.g. fallback from RNDV GET Zcopy (first time) to sending RTR (second time)

File: src/ucp/core/ucp_request.h (hunk 5 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_request.h]: removed it and all occurences

**Comment by dmitrygx** [comment]
bot:pipe:retest

**Comment by dmitrygx** [comment]
@brminich @yosefe could you review pls?

File: src/ucp/core/ucp_request.inl (hunk 6 above)
Comment (yosefe) [reply src/ucp/core/ucp_request.inl:930]: can we avoid 2nd set of super_req, or reset flags during the fallback?

File: src/ucp/core/ucp_request.h (hunk 5 above)
Comment (yosefe) [reply src/ucp/core/ucp_request.h]: i guess we can also undo ucp_request_reset_external now

File: src/ucp/core/ucp_request.inl (hunk 6 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_request.inl:930]: added reseting

File: src/ucp/core/ucp_request.h (hunk 5 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_request.h]: > i guess we can also undo ucp_request_reset_external now

done

**Comment by dmitrygx** [comment]
@yosefe @brminich your comments were applied
Could you pls review?

**Review by brminich — APPROVED** [review]

**Review by yosefe — APPROVED** [review]

**Pull request merged**